src/backend/distributed/utils/node_metadata.c

```diff
@@ -145,6 +148,112 @@ master_initialize_node_metadata(PG_FUNCTION_ARGS)
 
 
 /*
+ * get_shard_id_for_distribution_column function takes a distributed table name and a
+ * distribution value then returns shard id of the shard which belongs to given table and
+ * contains given value. This function only works for hash distributed tables.
+ */
+Datum
+get_shard_id_for_distribution_column(PG_FUNCTION_ARGS)
+{
+	Oid relationId = InvalidOid;
+	Datum distributionValue = 0;
+
+	Var *distributionColumn = NULL;
+	char distributionMethod = 0;
+	Oid expectedElementType = InvalidOid;
+	Oid inputElementType = InvalidOid;
+	DistTableCacheEntry *cacheEntry = NULL;
+	int shardCount = 0;
+	ShardInterval **shardIntervalArray = NULL;
+	FmgrInfo *hashFunction = NULL;
+	FmgrInfo *compareFunction = NULL;
+	bool useBinarySearch = true;
+	ShardInterval *shardInterval = NULL;
+
+	/*
+	 * To have optional parameter as NULL, we defined this UDF as not strict, therefore
+	 * we need to check all parameters for NULL values.
+	 */
+	if (PG_ARGISNULL(0))
+	{
+		ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+						errmsg("relation cannot be NULL")));
+	}
+
+	relationId = PG_GETARG_OID(0);
+	EnsureTablePermissions(relationId, ACL_SELECT);
+
+	if (!IsDistributedTable(relationId))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+						errmsg("relation is not distributed")));
+	}
+
+	distributionMethod = PartitionMethod(relationId);
+	if (distributionMethod == DISTRIBUTE_BY_NONE)
+	{
+		shardInterval = (ShardInterval *) linitial(LoadShardIntervalList(relationId));
```

Comment:
Minor: we'd probably crash here if there are no shards, might be cleaner to just return NULL for people with scripts that modify the metadata.